src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,1209 @@
+﻿// -----------------------------------------------------------------------
+//  <copyright file="ShardedSubscriptionConnection.cs" company="Hibernating Rhinos LTD">
+//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
+//  </copyright>
+// ----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Commands;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Subscriptions;
+using Raven.Client.Exceptions;
+using Raven.Client.Exceptions.Database;
+using Raven.Client.Exceptions.Documents.Subscriptions;
+using Raven.Client.Exceptions.Security;
+using Raven.Client.Http;
+using Raven.Client.Json.Serialization;
+using Raven.Client.ServerWide.Tcp;
+using Raven.Client.Util;
+using Raven.Server.Documents.Subscriptions;
+using Raven.Server.Documents.Subscriptions.Stats;
+using Raven.Server.Documents.TcpHandlers;
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Utils;
+// ReSharper disable AccessToDisposedClosure
+
+namespace Raven.Server.Documents.ShardedTcpHandlers
+{
+    public class ShardedSubscriptionConnection : SubscriptionConnectionBase, IDisposable
+    {
+        private List<SubscriptionServerWorker> _workers = new List<SubscriptionServerWorker>();
+        private HashSet<string> _rejectedShards = new HashSet<string>();
+        private readonly TimeSpan _rejectedReconnectTimeout = TimeSpan.FromMinutes(15); // TODO: egor create configuration
+        private bool _isDisposed;
+
+        private ShardedSubscriptionConnection(ServerStore serverStore, TcpConnectionOptions tcpConnectionOptions, IDisposable tcpConnectionDisposable, JsonOperationContext.MemoryBuffer buffer)
+            : base(tcpConnectionOptions, serverStore, buffer, tcpConnectionDisposable)
+        {
+        }
+
+        //TODO: egor handle double code with InitAsync and AssertConnection
+        public async Task TryReconnectRejectedShards()
+        {
+            var sp = new Stopwatch();
+            while (CancellationTokenSource.IsCancellationRequested == false)
+            {
+                if (_rejectedShards.Count == 0)
+                {
+                    await Task.Delay(1000);
+                    if(sp.IsRunning)
+                        sp.Reset();
+
+                    continue;
+                }
+
+                var rejectedShards = _rejectedShards;
+                if (CancellationTokenSource.IsCancellationRequested)
+                    break;
+
+                if (sp.IsRunning == false)
+                    sp.Start();
+      
```

Comment:
Use `DateTime.UtcNow` instead, we are talking about minutes differences here, after all. And the code is much more complex with stop watch